src/adc.rs

```diff
@@ -0,0 +1,319 @@
+}
+
+impl Channel<Adc<ADC>> for Pin<P5, Pin3, Alternate3<Input<Floating>>> {
+    type ID = u8;
+
+    fn channel() -> Self::ID { 11 }
+}
+
+pub struct Adc<ADC> {
+    adc_reg: ADC,
+}
+
+pub struct AdcConfig {
+    adc: ADC,
+    clock_source: ClockSource,
+    clock_divider: ClockDivider,
+    predivider: Predivider,
+    resolution: Resolution,
+    sampling_rate: SamplingRate,
+    sample_time: SampleTime,
+}
+
+impl AdcConfig {
+    pub fn new(
+        adc: ADC,
+        clock_source: ClockSource,
+        clock_divider: ClockDivider,
+        predivider: Predivider,
+        resolution: Resolution,
+        sampling_rate: SamplingRate,
+        sample_time: SampleTime,
+    ) -> AdcConfig {
+        AdcConfig {
+            adc,
+            clock_source: clock_source,
+            clock_divider: clock_divider,
+            predivider: predivider,
+            resolution: resolution,
+            sampling_rate: sampling_rate,
+            sample_time: sample_time,
+        }
+    }
+
+    pub fn config_hw(self) -> Adc<ADC> {
+        let adc_reg = self.adc;
+
+        adc_reg.adcctl0.modify(|_, w| w.adcenc().adcenc_0()
+                                                .adcon().adcon_0()
+                                                .adcsc().adcsc_0());
+
+        let adcsht = self.sample_time.adcsht();
+        adc_reg.adcctl0.modify(|_, w| w.adcsht().bits(adcsht));
+
+        let adcssel = self.clock_source.adcssel();
+        adc_reg.adcctl1.modify(|_, w| w.adcssel().bits(adcssel).adcshp().adcshp_1());
+
+        let adcdiv = self.clock_divider.adcdiv();
+        adc_reg.adcctl1.modify(|_, w| w.adcdiv().bits(adcdiv));
+
+        let adcpdiv = self.predivider.adcpdiv();
+        adc_reg.adcctl2.modify(|_, w| w.adcpdiv().bits(adcpdiv));
+
+        let adcres = self.resolution.adcres();
+        adc_reg.adcctl2.modify(|_, w| w.adcres().bits(adcres));
+
+        let adcsr = self.sampling_rate.adcsr();
+        adc_reg.adcctl2.modify(|_, w| w.adcsr().bit(adcsr));
+
+        Adc { adc_reg }
+    }
+}
+
+impl Adc<ADC> {
+    pub fn new(adc: ADC) -> Adc<ADC> {
+        Adc { adc_reg: adc }
+    }
+
+    pub fn adc_enable(&self) {
+        self.adc_reg.adcctl0.modify(|_, w| w.adcon().adcon_1());
+    }
+
+    pub fn adc_disable(&self) {
+        self.adc_reg.adcctl0.modify(|_, w| 
+            w.adcon().adcon_0()
+            .adcenc().adcenc_0());
+    }
+
+    pub fn adc_start_conversion(&self) {
+        self.adc_reg
+            .adcctl0
+            .modify(|_, w| w.adcenc().adcenc_1().adcsc().adcsc_1());
+    }
+
+    pub fn adc_is_busy(&self) -> bool {
+        return self.adc_reg.adcctl1.read().adcbusy().bit_is_set();
+    }
+
+    pub fn adc_get_result(&self) -> u16 {
+        return self.adc_reg.adcmem0.read().bits();
+    }
+
+    pub fn adc_set_pin<PIN>(&self, _pin: &PIN)
+    where PIN: Channel<Adc<ADC>, ID=u8> {
+        self.adc_reg.adcmctl0.modify(|_, w| w.adcinch().bits(PIN::channel()));
+    }

```

Comment:
APIs that use `modify` should either be changed to use `set_bits`/`clear_bits`/`toggle_bits` if possible or take `&mut self`.

Reply:
These public APIs should also be documented